File: cloudflare/crypto.js
// NovaPay V6.0 — Crypto utilities for Cloudflare Worker

let _jwtSecret = null;
let _aesKeyBytes = null;

export function setEnvSecrets(secretKey, aesKey) {
  _jwtSecret = secretKey || 'novapay-jwt-secret-change-in-production!';
  if (aesKey && aesKey.length >= 32) {
    _aesKeyBytes = new TextEncoder().encode(aesKey.slice(0, 32));
  }
}

async function getJwtSecret() {
  // Try to get from global env if available
  if (typeof globalThis !== 'undefined' && globalThis.env && globalThis.env.JWT_SECRET_KEY) {
    return globalThis.env.JWT_SECRET_KEY;
  }
  if (_jwtSecret) return _jwtSecret;
  return 'd7bbd897f4b41ac9da4f670492a8d6f8';
}

async function getAesKey() {
  const raw = _aesKeyBytes || 'c4902d50f71fe95acf32301ade5c3630';
  let keyBytes;
  if (raw.length >= 32) {
    keyBytes = new TextEncoder().encode(raw.slice(0, 32));
  } else if (raw.length > 0) {
    const hash = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(raw));
    keyBytes = new Uint8Array(hash);
  } else {
    keyBytes = new TextEncoder().encode('c4902d50f71fe95acf32301ade5c3630');
  }
  return crypto.subtle.importKey('raw', keyBytes, { name: 'AES-GCM' }, false, ['encrypt', 'decrypt']);
}

export async function aesEncrypt(plaintext) {
  if (!plaintext) return '';
  const key = await getAesKey();
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const encoded = new TextEncoder().encode(plaintext);
  const ciphertext = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, encoded);
  const ctBytes = new Uint8Array(ciphertext);
  const combined = new Uint8Array(12 + ctBytes.length);
  combined.set(iv);
  combined.set(ctBytes, 12);
  return btoa(String.fromCharCode(...combined));
}

export async function aesDecrypt(tokenB64) {
  if (!tokenB64) return '';
  const raw = atob(tokenB64);
  const bytes = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) bytes[i] = raw.charCodeAt(i);
  const iv = bytes.slice(0, 12);
  const ct = bytes.slice(12);
  const key = await getAesKey();
  const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, ct);
  return new TextDecoder().decode(decrypted);
}

export async function hashSecret(secret) {
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const key = await crypto.subtle.importKey('raw', new TextEncoder().encode(secret), 'PBKDF2', false, ['deriveBits']);
  const bits = await crypto.subtle.deriveBits(
    { name: 'PBKDF2', salt, iterations: 60000, hash: 'SHA-256' }, key, 256
  );
  const hashHex = Array.from(new Uint8Array(bits)).map(b => b.toString(16).padStart(2, '0')).join('');
  const saltHex = Array.from(salt).map(b => b.toString(16).padStart(2, '0')).join('');
  return `pbkdf2_sha256$60000$${saltHex}$${hashHex}`;
}

export async function verifySecret(hashStr, secret) {
  if (!hashStr || !secret) return false;
  try {
    const parts = hashStr.split('$');
    if (parts.length !== 4 || parts[0] !== 'pbkdf2_sha256') return false;
    const [, , saltHex, expectedHashHex] = parts;
    const salt = new Uint8Array(saltHex.match(/.{2}/g).map(b => parseInt(b, 16)));
    const key = await crypto.subtle.importKey('raw', new TextEncoder().encode(secret), 'PBKDF2', false, ['deriveBits']);
    const bits = await crypto.subtle.deriveBits(
      { name: 'PBKDF2', salt, iterations: 60000, hash: 'SHA-256' }, key, 256
    );
    const hashHex = Array.from(new Uint8Array(bits)).map(b => b.toString(16).padStart(2, '0')).join('');
    return hashHex === expectedHashHex;
  } catch (e) {
    return false;
  }
}

export async function sha256Hex(content) {
  const msgUint8 = new TextEncoder().encode(content || '');
  const hashBuffer = await crypto.subtle.digest('SHA-256', msgUint8);
  return Array.from(new Uint8Array(hashBuffer)).map(b => b.toString(16).padStart(2, '0')).join('');
}

export async function encodeJwt(payload) {
  const header = btoa(JSON.stringify({ alg: 'HS256', typ: 'JWT' })).replace(/\+/g, '-').replace(/\//g, '_').replace(/=$/g, '');
  const now = Math.floor(Date.now() / 1000);
  const claim = btoa(JSON.stringify({ ...payload, iat: now, exp: now + 3600 })).replace(/\+/g, '-').replace(/\//g, '_').replace(/=$/g, '');
  const sigInput = `${header}.${claim}`;
  const secretKey = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(await getJwtSecret()),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  const sigBytes = await crypto.subtle.sign('HMAC', secretKey, new TextEncoder().encode(sigInput));
  const sig = btoa(String.fromCharCode(...new Uint8Array(sigBytes))).replace(/\+/g, '-').replace(/\//g, '_').replace(/=$/g, '');
  return `${sigInput}.${sig}`;
}

export async function decodeJwt(token) {
  try {
    const [header, claim, sig] = token.split('.');
    if (!header || !claim || !sig) return null;
    // Convert base64url to base64 for atob (add padding)
    const b64url = s => {
      s = s.replace(/-/g, '+').replace(/_/g, '/');
      while (s.length % 4) s += '=';
      return s;
    };
    const payload = JSON.parse(atob(b64url(claim)));
    const sigInput = `${header}.${claim}`;
    const sigBytes = Uint8Array.from(atob(b64url(sig)), c => c.charCodeAt(0));
    const secretKey = await crypto.subtle.importKey(
      'raw',
      new TextEncoder().encode(await getJwtSecret()),
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['verify']
    );
    const verified = await crypto.subtle.verify('HMAC', secretKey, sigBytes, new TextEncoder().encode(sigInput));
    if (!verified) return null;
    if (payload.exp * 1000 < Date.now()) return null;
    return payload;
  } catch (e) {
    return null;
  }
}

export function nowMs() {
  const d = new Date();
  return d.toISOString().replace('T', ' ').slice(0, 23);
}

export function genUuid() {
  return crypto.randomUUID();
}

export function genTxId() {
  const ts = new Date().toISOString().replace(/[-:T]/g, '').slice(0, 14);
  const rand = Math.random().toString(36).slice(2, 8).toUpperCase();
  return 'TX' + ts + rand;
}

export function genGiftCode() {
  const chars = '0123456789';
  return Array.from({ length: 4 }, () => Array.from({ length: 4 }, () => chars[Math.floor(Math.random() * chars.length)]).join('')).join('-');
}

export function genUserId() {
  return 'NP' + String(Math.floor(Math.random() * 100000000)).padStart(8, '0');
}

File: app/crypto.py
"""
NovaPay V6.0 — 加密与安全工具
集中实现：
  - AES-256-GCM 对称加密（用于 CVV、用户建议内容）
  - Argon2id 密码 / PIN 哈希
  - SHA-256 内容防篡改哈希（用于建议完整性校验）
  - JWT 签发 / 校验（HTTP-only Cookie 配合）

所有密钥均来自 config（环境变量），本模块不硬编码任何密钥。
"""
import base64
import hashlib
import json
import time
from datetime import datetime, timezone

import jwt
from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError, InvalidHashError
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from config import AES_KEY, JWT_SECRET_KEY, JWT_ALGO, JWT_EXP_SECONDS

_ph = PasswordHasher()  # 默认 argon2id


# ----------------------------------------------------------------------------
# AES-256-GCM
# ----------------------------------------------------------------------------
def aes_encrypt(plaintext: str) -> str:
    """使用 AES-256-GCM 加密明文，返回 base64(nonce || tag || ciphertext)。"""
    if plaintext is None:
        plaintext = ""
    data = plaintext.encode("utf-8")
    aes = AESGCM(AES_KEY)
    nonce = os_urandom(12)
    ct = aes.encrypt(nonce, data, None)  # ct 末尾 16 字节为 GCM tag
    return base64.b64encode(nonce + ct).decode("ascii")


def aes_decrypt(token_b64: str) -> str:
    """解密 AES-256-GCM 密文，返回明文。校验失败抛出 InvalidTag。"""
    if not token_b64:
        return ""
    raw = base64.b64decode(token_b64)
    nonce, ct = raw[:12], raw[12:]
    aes = AESGCM(AES_KEY)
    pt = aes.decrypt(nonce, ct, None)
    return pt.decode("utf-8")


def os_urandom(n: int) -> bytes:
    import os
    return os.urandom(n)


# ----------------------------------------------------------------------------
# Argon2id 密码 / PIN
# ----------------------------------------------------------------------------
def hash_secret(secret: str) -> str:
    """对密码或 PIN 进行 Argon2id 哈希。"""
    return _ph.hash(secret)


def verify_secret(hash_str: str, secret: str) -> bool:
    """校验密码或 PIN。哈希非法或损坏时返回 False。"""
    if not hash_str:
        return False
    try:
        return _ph.verify(hash_str, secret)
    except (VerifyMismatchError, InvalidHashError):
        return False


def needs_rehash(hash_str: str) -> bool:
    try:
        return _ph.check_needs_rehash(hash_str)
    except Exception:
        return False


# ----------------------------------------------------------------------------
# SHA-256 防篡改哈希
# ----------------------------------------------------------------------------
def sha256_hex(content: str) -> str:
    """计算内容 SHA-256 十六进制摘要，用于建议完整性校验。"""
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


# ----------------------------------------------------------------------------
# JWT
# ----------------------------------------------------------------------------
def encode_jwt(payload: dict) -> str:
    """签发 JWT，附带过期时间与签发时间。"""
    now = int(time.time())
    body = dict(payload)
    body["iat"] = now
    body["exp"] = now + JWT_EXP_SECONDS
    return jwt.encode(body, JWT_SECRET_KEY, algorithm=JWT_ALGO)


def decode_jwt(token: str):
    """校验并解码 JWT，失败返回 None。"""
    try:
        return jwt.decode(token, JWT_SECRET_KEY, algorithms=[JWT_ALGO])
    except jwt.PyJWTError:
        return None


def now_ms() -> str:
    """返回精确到毫秒的 UTC 时间戳字符串（用于建议提交时间等）。"""
    return datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.") + f"{datetime.utcnow().microsecond // 1000:03d}"


def now_dt() -> datetime:
    return datetime.utcnow()
